sw/device/tests/sram_ctrl_main_scrambled_access_test.c

```diff
@@ -0,0 +1,177 @@
+
+/**
+ * MAIN SRAM scrambling test buffer.
+ *
+ * This buffer is initialized with a known pattern, which will be overwritten
+ * with a pseudo-random data as the result of MAIN SRAM scrambling.
+ */
+static volatile uint32_t
+    sram_ctrl_main_scramble_buffer[SRAM_CTRL_TESTUTILS_DATA_NUM_WORDS];
+
+/**
+ * Test pattern to be written and read from SRAM.
+ */
+static const sram_ctrl_testutils_data_t kRamTestPattern1 = {
+    .words =
+        {
+            0xA5A5A5A5,
+            0xA5A5A5A5,
+            0xA5A5A5A5,
+            0xA5A5A5A5,
+        },
+};
+
+/**
+ * Test pattern to be written and read from SRAM.
+ */
+static const sram_ctrl_testutils_data_t kRamTestPattern2 = {
+    .words =
+        {
+            0x5A5A5A5A,
+            0x5A5A5A5A,
+            0x5A5A5A5A,
+            0x5A5A5A5A,
+        },
+};
+
+/**
+ * Performs scrambling, saves the test relevant data and resets the system.
+ *
+ * MAIN SRAM will be overwritten with the pseudo-random data (including .bss
+ * and .data segments), which means that the code must be written in assembly.
+ *
+ * This function saves contents of the `sram_ctrl_main_scramble_buffer` in the
+ * RETENTION SRAM, which maintains the power across the system reboot.
+ */
+static noreturn void main_sram_scramble(void) {
+  asm volatile(
+      // Request the new scrambling key for MAIN SRAM.
+      "li %[kGeneralRW], 0x1;                                              \n"
+      "sw %[kGeneralRW], %[kSramCtrlOffset](%[kSramCtrlRegsBase]);         \n"
+
+      // Busy loop - waiting for scrambling to finish.
+      ".L_SCRAMBLING_BUSY_LOOP:;                                           \n"
+      "  lw %[kGeneralRW], %[kSramCtrlStatusOffset](%[kSramCtrlRegsBase]); \n"
+      "  andi %[kGeneralRW], %[kGeneralRW], %[kSramCtrlKeyScrDone];        \n"
+      "  beqz %[kGeneralRW], .L_SCRAMBLING_BUSY_LOOP;                      \n"
+
+      // Copy the buffer contents to the retention SRAM.
+      ".L_BUFFER_COPY_LOOP:;                                               \n"
+      "  lw %[kGeneralRW], 0(%[kCopyFromAddr]);                            \n"
+      "  sw %[kGeneralRW], 0(%[kCopyToAddr]);                              \n"
+      "  addi %[kCopyFromAddr], %[kCopyFromAddr], 4;                       \n"
+      "  addi %[kCopyToAddr], %[kCopyToAddr], 4;                           \n"
+      "  blt %[kCopyToAddr], %[kCopyToEndAddr], .L_BUFFER_COPY_LOOP;       \n"
+
+      // Trigger the software system reset via the Reset Manager.
+      "li %[kGeneralRW], %[kMultiBitTrue];                                 \n"
+      "sw %[kGeneralRW], %[kRstMgrResetReq](%[kRstMgrRegsBase]);           \n"
+
+      // Satisfy the `noreturn` promise to the compiler.
+      ".L_INFINITE_LOOP:;                                                  \n"
+      "  wfi;                                                              \n"
+      "  j .L_INFINITE_LOOP;"
+      :
+      : [kGeneralRW] "r"(0), [kMultiBitTrue] "I"(kMultiBitBool4True),
```

Comment:
Can you rename `kGeneralRW` to `tmp` and maybe add a comment that it is allocating a register for general use? It's not a constant so we should probably not use the `k` prefix.

side note: clang format makes some strange decisions with inline assembly, sigh